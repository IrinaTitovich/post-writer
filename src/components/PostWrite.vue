<script lang="ts" setup>
import { onMounted, ref, watch } from "vue";
import { marked } from "marked";
import highlightsjs from "highlight.js";
import lodash from "lodash";
import { Post, TimeLinePost } from "../posts";
import { useUsers } from "../stores/userStore";

const props = defineProps<{
  post: TimeLinePost | Post;
}>();

const emit = defineEmits<{
  (event: "submit", post: Post): void;
}>();

const title = ref(props.post.title);
const content = ref(props.post.markdown);
const html = ref("");
const contentEditable = ref<HTMLDivElement>();

onMounted(() => {
  if (!contentEditable.value) {
    throw new Error("Content Editable is not found");
  }
  contentEditable.value.innerText = content.value;
});

const onInput = () => {
  if (!contentEditable.value) {
    throw new Error("Content Editable is not found");
  }

  content.value = contentEditable.value.innerText;
};

const updateHtml = (value: string) => {
  marked.parse(
    value,
    {
      gfm: true,
      breaks: true,
      highlight: (code) => {
        return highlightsjs.highlightAuto(code).value;
      },
    },
    (_, parseResult) => {
      html.value = parseResult;
    }
  );
};

watch(
  content,
  lodash.debounce((contentValue) => updateHtml(contentValue), 300),
  { immediate: true }
);

const userStore = useUsers();

function onSaveChanges() {
  if (!userStore.currentUserId) {
    throw Error("user was not found");
  }

  const post: Post = {
    ...props.post,
    created:
      typeof props.post.created === "string"
        ? props.post.created
        : props.post.created.toISODate(),
    title: title.value,
    authorId: userStore.currentUserId,
    html: html.value,
    markdown: content.value,
  };

  emit("submit", post);
}
</script>

<template>
  <div class="columns">
    <div class="column">
      <div class="field">
        <div class="label">{{ post.title }}</div>
        <input type="text" class="input" v-model="title" />
      </div>
    </div>
  </div>

  <div class="columns">
    <div class="column">
      <div
        id="contenteditable"
        contenteditable
        ref="contentEditable"
        @input="onInput"
      />
    </div>

    <div class="column">
      <div v-html="html" />
    </div>
  </div>
  <div class="columns">
    <div class="column">
      <button class="button is-primary is-pulled-right" @click="onSaveChanges">
        Save Post
      </button>
    </div>
  </div>
</template>

<style>
@import "highlight.js/styles/atom-one-dark.css";
</style>
