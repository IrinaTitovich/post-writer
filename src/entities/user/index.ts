export { useUsers } from "./store/userStore";
export * from "./user";
export * from "./composables/modal";
