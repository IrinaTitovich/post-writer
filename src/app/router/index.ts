import { createRouter, createWebHistory } from "vue-router";
import { useUsers } from "../../entities/user";
import { EditPost, NewPost, PostView, ShowPost } from "../../pages";

export const routes = [
  {
    path: "/",
    component: PostView,
  },
  {
    path: "/posts/new",
    component: NewPost,
    beforeEnter: () => {
      const usersStore = useUsers();

      if (!usersStore.currentUserId) {
        return {
          path: "/",
        };
      }
    },
  },
  {
    path: "/posts/:id",
    component: ShowPost,
    beforeEnter: () => {
      const usersStore = useUsers();

      if (!usersStore.currentUserId) {
        return {
          path: "/",
        };
      }
    },
  },
  {
    path: "/posts/:id/edit",
    component: EditPost,
    beforeEnter: () => {
      const usersStore = useUsers();

      if (!usersStore.currentUserId) {
        return {
          path: "/",
        };
      }
    },
  },
];

export const router = createRouter({
  history: createWebHistory(),
  routes,
});
