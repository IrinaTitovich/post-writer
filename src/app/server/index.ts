import express, { Response } from "express";
import cors from "cors";
import bodyParser from "body-parser";
import cookieParser from "cookie-parser";
import jsonwebtoken from "jsonwebtoken";
import { today, thisMonth, thisWeek, Post } from "../../posts";
import { NewUser, User } from "../../user";

const app = express();
app.use(cors());
app.use(cookieParser());
app.use(bodyParser.json());

const posts = [today, thisWeek, thisMonth];
const users: User[] = [];

const SECRET = "my-secret";
const COOKIE = "jwt";

app.post<unknown, unknown, Post>("/posts", (req, res) => {
  const post = { ...req.body, id: (Math.random() * 10000).toFixed() };
  posts.push(post);

  res.json(post);
});

app.put<unknown, unknown, Post>("/posts", (req, res) => {
  const updatedPost = { ...req.body };
  const index = posts.findIndex(({ id }) => id === updatedPost.id);

  posts[index] = updatedPost;

  res.json(posts[index]);
});

function autentification(id: string, res: Response) {
  const token = jsonwebtoken.sign({ id }, SECRET, {
    issuer: "IT",
    expiresIn: "30 days",
  });

  res.cookie(COOKIE, token, { httpOnly: true });
}

app.get("/current-user", (req, res) => {
  try {
    const token = req.cookies[COOKIE];
    const result = jsonwebtoken.verify(token, SECRET) as { id: string };

    res.json({ id: result.id });
  } catch (e) {
    res.status(404).end();
  }
});

app.get("/posts", (req, res) => {
  res.json(posts);
});

app.post<unknown, unknown, NewUser>("/users", (req, res) => {
  const user: User = { ...req.body, id: (Math.random() * 10000).toFixed() };
  users.push(user);
  autentification(user.id, res);

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { password, ...userResponse } = user;

  res.json({ userResponse });
});

app.post<unknown, unknown, NewUser>("/login", (req, res) => {
  const user = { ...req.body };

  const existingUser = users.find((u) => u.username === user.username);

  if (existingUser && user.password === existingUser.password) {
    autentification(existingUser.id, res);
    res.status(200).end();
  } else {
    res.status(401).end();
  }
});

app.post("/logout", (req, res) => {
  res.cookie(COOKIE, "", { httpOnly: true });
  res.status(200).end();
});

app.listen(8000, () => {
  console.log("Listening on port 8000");
});
