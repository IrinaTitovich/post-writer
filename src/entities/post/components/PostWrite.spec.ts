import { mount } from "@vue/test-utils";
import { createPinia, Pinia, setActivePinia } from "pinia";
import { describe, beforeEach, it, expect } from "vitest";
import { VueQueryPlugin } from "@tanstack/vue-query";
import { createRouter, createWebHistory, Router } from "vue-router";
import { routes } from "../../../app/router";
import { useUsers } from "../../user";
import PostWrite from "./PostWrite.vue";

describe("PostWrite", () => {
  let pinia: Pinia;
  let router: Router;
  const CURRENT_USER = "1";

  beforeEach(() => {
    pinia = createPinia();
    setActivePinia(pinia);

    const userStore = useUsers();
    userStore.currentUserId = CURRENT_USER;

    router = createRouter({
      history: createWebHistory(),
      routes,
    });
  });

  it("renders a post using markdown", async () => {
    const wrapper = mount(PostWrite, {
      global: {
        plugins: [pinia, router, VueQueryPlugin],
      },
      props: {
        post: {
          id: "145",
          title: "",
          created: "",
          authorId: CURRENT_USER,
          markdown: "",
          html: "",
        },
      },
    });

    const input = wrapper.find<HTMLDivElement>("#contenteditable");

    input.element.innerHTML = "# Title";

    await input.trigger("input");

    expect(wrapper.text()).toContain("Title");
  });

  it("emits a post correctly", () => {
    return new Promise<void>((resolve) => {
      const wrapper = mount(PostWrite, {
        global: {
          plugins: [pinia, router],
        },
        props: {
          post: {
            id: "145",
            title: "",
            created: "",
            authorId: CURRENT_USER,
            markdown: "",
            html: "",
          },
        },
      });

      const input = wrapper.find<HTMLDivElement>("#contenteditable");

      input.element.innerHTML = "# Title";
      input.trigger("input");

      setTimeout(async () => {
        await wrapper.find("button").trigger("click");

        expect(wrapper.emitted().submit[0]).toMatchInlineSnapshot(`
          [
            {
              "authorId": "1",
              "created": "",
              "html": "",
              "id": "145",
              "markdown": "",
              "title": "",
            },
          ]
        `);

        expect(wrapper.emitted().submit[0]).toStrictEqual([
          {
            id: "145",
            title: "",
            created: "",
            authorId: "1",
            markdown: "",
            html: "",
          },
        ]);

        resolve();
      }, 300);
    });
  });
});
