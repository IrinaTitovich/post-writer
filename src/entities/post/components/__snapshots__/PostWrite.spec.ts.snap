// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostWrite > writes a post using markdown 1`] = `
[
  {
    "authorId": "1",
    "created": "",
    "html": "",
    "id": "145",
    "markdown": "",
    "title": "",
  },
]
`;
